include/ogdf/basic/internal/copy_move.h

```diff
@@ -0,0 +1,90 @@
+ * Version 2 or 3 as published by the Free Software Foundation;
+ * see the file LICENSE.txt included in the packaging of this file
+ * for details.
+ *
+ * \par
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * \par
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, see
+ * http://www.gnu.org/copyleft/gpl.html
+ */
+
+#pragma once
+
+/* This needs to be synchronized with doc/ogdf-doxygen.cfg:EXPAND_AS_DEFINED */
+
+//! Explicitly disables (deletes) copy construction and assignment for class \p cls.
+#define OGDF_NO_COPY(cls)          \
+	cls(const cls& copy) = delete; \
+	cls& operator=(const cls& copy) = delete;
+
+//! Explicitly disables (deletes) move construction and assignment for class \p cls.
+#define OGDF_NO_MOVE(cls)     \
+	cls(cls&& move) = delete; \
+	cls& operator=(cls&& move) = delete;
+
+//! Explicitly provides default copy construction and assignment for class \p cls.
+#define OGDF_DEFAULT_COPY(cls)      \
+	cls(const cls& copy) = default; \
+	cls& operator=(const cls& copy) = default;
+
+//! Explicitly provides default move construction and assignment for class \p cls.
+#define OGDF_DEFAULT_MOVE(cls) \
+	cls(cls&& move) = default; \
+	cls& operator=(cls&& move) = default;
+
+//! Declares the copy constructor for class \p cls.
+#define OGDF_COPY_CONSTR(cls) cls(const cls& copy)
+//! Declares the copy assignment operation for class \p cls.
+#define OGDF_COPY_OP(cls) cls& operator=(const cls& copy)
+//! Declares the move constructor for class \p cls.
+#define OGDF_MOVE_CONSTR(cls) cls(cls&& move) noexcept : cls()
+//! Declares the move assignment operation for class \p cls.
+#define OGDF_MOVE_OP(cls) cls& operator=(cls&& move) noexcept
+//! Declares the swap function for class \p cls.
+#define OGDF_SWAP_OP(cls) friend void swap(cls& first, cls& second) noexcept
+
+//! Provide move construct/assign and copy assign given there is a copy constructor and swap.
+/**
+ * Only requires custom implementations of OGDF_COPY_CONSTR and OGDF_SWAP_OP
+ * and automatically provides OGDF_COPY_OP, OGDF_MOVE_CONSTR, and OGDF_MOVE_OP.
+ *
+ * See https://stackoverflow.com/a/3279550 for more details on this idiom.
+ */
+#define OGDF_COPY_MOVE_BY_SWAP(cls)                       \
+	/*! Assign this cls to be a copy of \p copy_by_value.
+	    Internally, this will use the copy constructor to create a temporary copy and then swap
+	    this object instance with the temporary copy using via a std::swap implementation.
+	    See https://stackoverflow.com/a/3279550 for more details on this idiom. */ \
+	cls& operator=(cls copy_by_value) {                   \
+		using std::swap;                                  \
+		swap(*this, copy_by_value);                       \
```

Comment:
OGDF_COPY_MOVE_BY_SWAP is never actually used, so it's difficult to see whether it actually works (and whether the documentation for it is built correctly). I assume its intention is to be used for classes that should be supported by e.g. NodeArray/EdgeArray...?

Right now I already see the following problems:
* This `operator=` uses `std::swap` but should actually use a user-defined one (here: the OGDF_SWAP_OP one). The stackoverflow post mentioned above explicitly states:
"We might be tempted to use std::swap instead of providing our own, but this would be impossible; std::swap uses the copy-constructor and copy-assignment operator within its implementation, and we'd ultimately be trying to define the assignment operator in terms of itself!"
* You need `return *this;` at the of this function. You also need it in OGDF_MOVE_OP below.

If we add this macro, we should at least do a manual test once to see that it works for a dummy class (even if we don't have an explicit test case for it right now).

Reply:
I fixed your points and made `DisjointSets` copyable with this approach. What occurred to me is that the `operator=` here might get a `noexcept` declaration if the selected copy or move constructor and swap implementations allow. Will investigate how to do this...

Reply:
Now correctly marked as `noexcept` and the docs look good, so this should be good to merge.

Reply:
Nice! Thank you for your work, I'll merge :turtle: